File: python/src/patterns/cqrs/event_bus.py
from typing import Callable
from .events import DomainEvent

Handler = Callable[[DomainEvent], None]

class EventBus:
    def __init__(self) -> None:
        self._handlers: list[Handler] = []
    def publish(self, e: DomainEvent) -> None:
        for h in list(self._handlers):
            h(e)
    def subscribe(self, h: Handler):
        self._handlers.append(h)
        def unsubscribe():
            self._handlers.remove(h)
        return unsubscribe

File: typescript/src/patterns/cqrs/EventBus.ts
import type { DomainEvent } from './Events.js';
export type Handler = (e: DomainEvent) => void;

export class EventBus {
  private handlers: Handler[] = [];
  publish(e: DomainEvent) { this.handlers.forEach(h => h(e)); }
  subscribe(h: Handler) { this.handlers.push(h); return () => this.handlers = this.handlers.filter(x => x !== h); }
}
